velox/functions/prestosql/types/IPPrefixType.cpp

```diff
@@ -0,0 +1,326 @@
+      case TypeKind::VARCHAR:
+        return true;
+      case TypeKind::HUGEINT:
+        if (isIPAddressType(other)) {
+          return true;
+        }
+      default:
+        return false;
+    }
+  }
+
+  bool isSupportedToType(const TypePtr& other) const override {
+    switch (other->kind()) {
+      case TypeKind::VARCHAR:
+        return true;
+      case TypeKind::HUGEINT:
+        if (isIPAddressType(other)) {
+          return true;
+        }
+      default:
+        return false;
+    }
+  }
+
+  void castTo(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (input.typeKind() == TypeKind::VARCHAR) {
+      castFromString(input, context, rows, *result);
+    } else if (isIPAddressType(input.type())) {
+      castFromIPAddress(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from {} to IPPrefix not yet supported",
+          input.type()->toString());
+    }
+  }
+
+  void castFrom(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, resultType, result);
+
+    if (resultType->kind() == TypeKind::VARCHAR) {
+      castToString(input, context, rows, *result);
+    } else if (isIPAddressType(resultType)) {
+      castToIPAddress(input, context, rows, *result);
+    } else {
+      VELOX_UNSUPPORTED(
+          "Cast from IPPrefix to {} not yet supported", resultType->toString());
+    }
+  }
+
+ private:
+  static void castToString(
+      const BaseVector& input,
+      exec::EvalCtx& context,
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipaddresses = input.as<SimpleVector<StringView>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto intAddr = ipaddresses->valueAt(row);
+      folly::ByteArray16 addrBytes;
+
+      memcpy(&addrBytes, intAddr.data(), kIPAddressBytes);
+      folly::IPAddressV6 v6Addr(addrBytes);
+
+      exec::StringWriter<false> result(flatResult, row);
+      if (v6Addr.isIPv4Mapped()) {
+        result.append(fmt::format(
+            "{}/{}",
+            v6Addr.createIPv4().str(),
+            (uint8_t)intAddr.data()[kIPAddressBytes]));
+      } else {
+        result.append(fmt::format(
+            "{}/{}", v6Addr.str(), (uint8_t)intAddr.data()[kIPAddressBytes]));
+      }
+      result.finalize();
+    });
+  }
+
+  static folly::small_vector<folly::StringPiece, 2> splitIpSlashCidr(
+      const folly::StringPiece& ipSlashCidr) {
+    folly::small_vector<folly::StringPiece, 2> vec;
+    folly::split('/', ipSlashCidr, vec);
+    return vec;
+  }
+
+  static void castFromString(
```

Comment:
Is this logic consistent with Java ? Would the same invalid values return the same errors ?

Reply:
Not the same error message, but will return an error for the same input as java.

Reply:
I agree that it might not be possible to use the same errors in all places. But we should be as consistent as we can else its very hard during any kind of traffic migration to validate the builds. Meta has big issues with it.

In general, I feel it might be good to split this work as follows:
i) First PR just adds all the type classes for the IPPrefixType. So this PR minus all the casting code.
ii) Add the fuzzer for IPPrefix Casts that compares with Presto.
iii) Add the cast code with the fuzzer in place. That will ensure that we have most consistency. This would also make it simpler to validate the code you have in IPPrefixCastTest.

Based on prior experience I feel the fuzzer should not be an afterthought.

@amitkdutta @kagamiori @kgpai

Reply:
i) https://github.com/facebookincubator/velox/pull/11122
ii) https://github.com/facebookincubator/velox/pull/11209